cilium, ipvlan: avoid tail call in fast path

borkmann opened the pull request:
(Mainly in context of ipvlan to reduce latency.)

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/6748)
<!-- Reviewable:end -->

borkmann committed:
cilium, bpf: explicitly set defines to 1

This is needed as it looks like clang has different behavior as
gcc on this regard wrt upcoming macro hackery for tail calls.

We should make it consistent for all ifdef type checks in general
though, right now it's a bit of a mixture in the code emission.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, bpf: config depdendent tail call emission in fast path

Tail calls are expensive due to being an indirect call and thus subject
to speculative execution which is mitigated these days via retpoline
in JIT. Lets add some macros that can be used with is_defined(...) to
switch between tail call and more performant direct inlined call for
fast-path. Given verifier complexity limits we need to be selective when
to enable. For now, go for avoiding tail calls when either v4 or v6 is
set but not both as it has highest complexity. x86-64 JIT retpoline
code-gen output:

  https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=a493a87f38cfa48caaa95c9347be2d914c6fdf29

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: unquote MAX_LXC_OPTIONS in build test

We need to unquote it as otherwise this doesn't compile when used
in macros. Unquoted they are in any case properly passed to clang.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

joestringer commented on bpf/lib/tailcall.h:
```diff
@@ -0,0 +1,39 @@
+		ep_tail_call(skb, NAME);      \
+		ret = DROP_MISSED_TAIL_CALL;  \
+	} while (0)
+#define invoke_tailcall_if(COND, NAME, FUNC)  \
```
I guess this is "invoke, or tail call if the condition COND is met" - it's admittedly getting longer but it'd be nice to make this clearer by either naming it `invoke_or_tailcall_if(COND...`, and/or adding a comment above this definition to describe it in more detail.

borkmann replied:
Will add a comment and integrate into policy bits too.

joestringer reviewed (approved):
I guess we could apply the same logic on ingress too? It was split in commit 7a98029b6b2cbc3a026d48ab2cf3637e53a7d985 but we were always enabling IPv6 at that time too.

Approving assuming that this doesn't break complexity on 4.9 kernels.

borkmann committed:
cilium, ipvlan: drop masquerade restriction

In Cilium ipvlan datapath mode, it can be enabled just fine since
there are two devices ipvlan master and cilium_host. Traffic leaving
the node via ipvlan master gets masqueraded with --masquerade=true
set but not traffic leaving cilium_host. Traffic back into the
latter is only from local traffic but not external hence the extra
-DENABLE_EXTRA_HOST_DEV for bpf_netdev attached there on egress in
order to match HOST_ID.

For the ipvlan master device, the external traffic is pushed right
into one of the slaves. The bpf_netdev on ingress comes before
reverse address translation on masquerade, but the bpf_netdev.o has
neither -DFROM_HOST nor -DENABLE_EXTRA_HOST_DEV set, so secctx will
be taken from packet if available (fallback to WORLD_ID e.g. in IPv4)
or from local ipcache if we only found WORLD_ID. For the latter
however, we only derive the secctx if it was not found to be HOST_ID
given masquerade.

Note, ipvlan l3s mode requires the following kernel commit in order
to work properly (it's being queued for stable kernel's as well):

  https://git.kernel.org/pub/scm/linux/kernel/git/davem/net.git/commit/?id=d5256083f62e2720f75bb3c5a928a0afe47d6bc3

With that applied, --masquerade=true works fine.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>
